pkg/storaged/client.js

```diff
@@ -206,6 +214,145 @@
+            } catch (err) {
+                console.error(`unable to obtain subvolumes for mount point ${mount_point}`, err);
+                return;
+            }
+            const subvols = [{ pathname: "/", id: 5 }];
+            for (const line of output.split("\n")) {
+                const m = line.match(/ID (\d+).*parent (\d+).*path (<FS_TREE>\/)?(.*)/);
+                if (m)
+                    subvols.push({ pathname: m[4], id: Number(m[1]) });
+            }
+            uuids_subvols[uuid] = subvols;
+
+            // HACK: UDisks should expose a better btrfs API with btrfs device information
+            // https://github.com/storaged-project/udisks/issues/1232
+            // TODO: optimise into just parsing one `btrfs filesystem show`?
+            const usages = {};
+            let usage_output;
+            try {
+                usage_output = await cockpit.spawn(["btrfs", "filesystem", "show", "--raw", uuid], { superuser: true, err: "message" });
+            } catch (err) {
+                console.error(`btrfs filesystem show ${uuid}`, err);
+                return;
+            }
+            for (const line of usage_output.split("\n")) {
+                const match = usage_regex.exec(line);
+                if (match) {
+                    const { used, device } = match.groups;
+                    usages[device] = used;
+                }
+            }
+            uuids_usage[uuid] = usages;
+        } else {
+            uuids_subvols[uuid] = null;
+            uuids_usage[uuid] = null;
+        }
+    }
+
+    // HACK: we should try to use --poll and run this separate from the 5 seconds btrfs subvolume poll
+    // This fails when no btrfs filesystem is found with the --mtab option and exits with 1, so that is kinda useless, however without --mtab
+    // we don't get a nice flat structure. So we ignore the errors
+    if (Object.keys(client.uuids_btrfs_volume).length > 0) {
+        let findmnt_output;
+        try {
+            findmnt_output = await cockpit.spawn(["findmnt", "--type", "btrfs", "--mtab", "-o", "UUID,OPTIONS,SOURCE,TARGET", "--json"],
+                                                 { superuser: true, err: "message" });
+        } catch (err) {
+            if (err.message)
+                console.error("error when running findmnt", err);
+            return;
+        }
+        try {
+            // Extract the data into a { uuid: { subvolid: { subvol, target } } }
+            const mounts = JSON.parse(findmnt_output);
+            if ("filesystems" in mounts) {
+                for (const fs of mounts.filesystems) {
+                    const subvolid_match = fs.options.match(/subvolid=(?<subvolid>\d+)/);
+                    const subvol_match = fs.options.match(/subvol=(?<subvol>[\w\\/]+)/);
+                    if (!subvolid_match && !subvol_match) {
+                        console.warn("findmnt entry without subvol and subvolid", fs);
+                        break;
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19690-20231208-171735-1465bfd6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)